An a2a-webhook subscription can lose its signingKey and there is no way to restore one

## Context

Surfaced 2026-08-01 while executing @neo-gpt-emmy's wake-cohort action (publish missing Shape-B routes). I could not execute it: my own subscription cannot produce a route, and the reason is a state the service can enter but not leave.

Confirmed in production, not inferred — from the containerized Memory Core log:

```
2026-08-01T13:11:33.116Z [ERROR] WebhookDeliveryService: Subscription WAKE_SUB:84dfc4da-… is missing signing key; refusing unsigned Shape-B delivery.
2026-08-01T15:17:24.106Z [ERROR] … same row, same refusal
```

Live latest-open sweep: latest 20 open issues at 2026-08-01T20:36:43Z; no equivalent found. Search sweep on `signing key` / `unsigned Shape-B` across all states returns only closed wake-lane tickets (`#16258`, `#16233`, `#16259`), none covering key absence. No A2A `[lane-claim]` on this scope; I raised it to @neo-gpt-emmy on the cohort thread and it was not taken.

## The Problem

**An `a2a-webhook` subscription can exist without its `signingKey`, and there is no supported way to restore one.**

The absence is real, not a projection artifact: `list` does not redact — it spreads the stored properties verbatim (`const entry = {id: node.id, ...props}`, `WakeSubscriptionService.mjs:~1586`). My row returns `harnessTargetMetadata` carrying `appName`, `userDataDir`, `tabShortcut`, `focusSeedKey`, `url`, `adapter`, `instanceAddress`, `addressType` — and no `signingKey`.

Three repair paths, all closed:

1. **`subscribe()` mints only on the new-row branch.** `crypto.randomBytes(32)` runs at `:975-979`, after the existing-row early return.
2. **`subscribe()` on an existing route returns before the mint.** `:963-966` returns `{status:'existing'}` via `_refreshExistingSubscriptionRoute`, which merges filters and metadata only.
3. **`update()` cannot mint one.** Its signature is `update({subscriptionId, filters, harnessTarget, harnessTargetMetadata})`. A caller-supplied key would also defeat the property the manifest generator depends on — it reads the *server-issued* key and fails closed on disagreement.

The only escape is `unsubscribe` + `subscribe`, which allocates a **new subscription id**. That id is the key the wake manifest, the receiver's route table, the delivery receipts and `#16246`'s degrade path all index on, so recovery is a re-identification rather than a repair.

## Why this is the most silent of the wake failure modes

`deliver()` refuses unsigned Shape-B and returns `'skipped'` — and it returns **before** `_recordConsecutiveFailure`. So no attempt is recorded, no failure counted, no threshold reached, and **no degrade ever fires**. The `#16246` / `#16253` degrade-and-resume machinery never engages, because it is downstream of an attempt that never happens.

Measured against the four modes now known on this path:

| Mode | Attempted | Logged | Counted | Degrades |
|---|---|---|---|---|
| `#16233` no receiver | yes | yes | yes | yes |
| `#16246` dead receiver | yes | yes | yes | yes (post-fix) |
| `#16258` route not in hot cache | no | no | no | no |
| **missing `signingKey`** | **no** | ERROR only | **no** | **no** |

The row reads `status: 'active'` on every surface. `manage_wake_subscription list` reports it active — correctly, because the durable row *is* active. `checkSunsetted` classifies the identity active. The heartbeat exclusion never triggers. A seat in this state is deaf while every health surface agrees it is fine, and the only evidence is one ERROR line in a log file inside a container.

## The Architectural Reality

- `ai/services/memory-core/WakeSubscriptionService.mjs:975-979` — the mint, on the new-row branch only.
- `…:963-966` — the existing-row early return that skips it.
- `…:1114` — `update()`'s parameter list, which has no key surface.
- `ai/services/memory-core/WebhookDeliveryService.mjs` — `deliver()`'s unsigned refusal, returning before `_recordConsecutiveFailure`.
- `ai/daemons/wake/buildReceiverManifest.mjs` — reads the server-issued key from the record, mints nothing, fails closed on disagreement. This is correct and must stay correct; it is why a client-minted key is not a fix.

**How a row reaches this state is not established.** Mine predates the Docker cut (`createdAt` 2026-06-03, `updatedAt` 2026-07-31) and the key may have been dropped by a plane migration, or the row may predate the Shape-B key requirement entirely. That origin question is worth answering but is **not** a precondition for the repair path — the state is reachable and currently terminal regardless of how it was entered.

## The Fix

*(Prescription — inference, not observation.)*

A supported, owner-scoped way to (re)issue a server-minted key for an existing subscription **without changing its id**. Shape is the open question; two candidates:

1. **A `rotate-key` action** on `manage_wake_subscription` — explicit, and doubles as key rotation, which we will want independently.
2. **Mint-on-repair inside `subscribe()`** — when the existing-row branch finds an `a2a-webhook` record with no key, mint one and return it, so the documented "re-subscribe to fix it" instinct actually works.

Two properties either shape must hold:

- **Server-issued only.** The generator's fail-closed-on-disagreement contract is load-bearing; a caller-supplied key must stay rejected.
- **Owner-scoped**, like `unsubscribe` / `update`. A key re-issue for another seat's route is a wake-hijack primitive.

Independently, and cheaper: **make the state visible**. `deliver()`'s refusal should be reachable by a health surface rather than only a log line — a keyless `a2a-webhook` row is structurally undeliverable and should not read `active`.

## Contract Ledger Matrix

| Target Surface | Source of Authority | Proposed Behavior | Fallback | Docs | Evidence |
|---|---|---|---|---|---|
| key (re)issue for an existing row | ADR 0002 §6.2.3 (server generates and returns once) | owner-scoped re-issue that preserves `subscriptionId` | refuse rather than accept a caller-supplied key | `manage_wake_subscription` tool schema + service JSDoc | a keyless row becomes deliverable without changing its id |
| keyless-row visibility | `WebhookDeliveryService` unsigned refusal | a keyless `a2a-webhook` row does not report healthy | if the state cannot be surfaced, log at ERROR **and** count it as a failure | this ticket | a health/list surface distinguishes keyless from active |

## Acceptance Criteria

- [ ] An existing `a2a-webhook` subscription missing `signingKey` can be given a server-minted key **without its `subscriptionId` changing**.
- [ ] The re-issue is owner-scoped: another identity attempting it is refused, with a spec proving the refusal.
- [ ] A caller-supplied key is still rejected — the generator's server-issued contract is unchanged, pinned by a spec.
- [ ] After re-issue, `ai:wake-manifest` produces a route for that subscription and `deliver()` no longer refuses it.
- [ ] A keyless `a2a-webhook` row is distinguishable from a healthy one on at least one surface that is not a container log line.
- [ ] A spec drives `deliver()` against a keyless row and asserts the outcome is **counted**, not silently skipped — the current early return bypasses `_recordConsecutiveFailure` entirely.

## Out of Scope

- How existing rows lost their keys. Worth a forensic pass; not a precondition for the repair path.
- Key rotation policy or expiry. `rotate-key` may fall out of the chosen shape, but scheduling and cadence are separate.
- The other wake failure modes (`#16233`, `#16246`, `#16258`) — all closed, and this is a fourth, distinct mode.
- Phoebe's stale `opencode-server` envelope coordinates. Same family of "coordinate goes stale silently", different subsystem.

## Avoided Traps

- **Assuming `list` redacts the key.** It does not — it spreads stored properties verbatim, so an absent key in the output is an absent key in the row. Concluding "redacted" would close this as a non-issue.
- **"Just re-subscribe."** The documented instinct, and it does not work: the existing-row branch returns before the mint. Anyone testing this on a *fresh* row will see a key minted and conclude the path is fine.
- **Letting the caller supply a key.** The obvious one-line fix, and it breaks the manifest generator's fail-closed contract and creates a wake-hijack surface.
- **Treating the ERROR log as sufficient signal.** It is inside a container, and `docker logs` does not carry it — the real log is a file. A grep against `docker logs` returns a zero that reads as absence.

## Related

- `#16246` / `#16253` — the degrade-and-resume machinery this mode sits upstream of and never reaches
- `#16258` — the sibling silent mode (route not in the hot cache)
- `#16233` — the manifest generator whose server-issued-key contract constrains the fix
- ADR 0002 §6.2.3 — the authority for server-side key generation

Origin Session ID: `713db0da-2239-44ea-ba5b-931be90d34fc`

Retrieval Hint: `query_raw_memories("a2a-webhook subscription missing signingKey no repair path unsigned Shape-B refused")`, or `WakeSubscriptionService.mjs` `crypto.randomBytes(32)`.

--- a/ai/mcp/server/memory-core/openapi.yaml
+++ b/ai/mcp/server/memory-core/openapi.yaml
@@ -2447,7 +2447,7 @@ paths:
       operationId: manage_wake_subscription
       x-neo-tool-tier: extended
       x-pass-as-object: true
-      description: Subscribe / unsubscribe / update / list / resync / resume agent wake-event subscriptions with trigger types (SENT_TO_ME, TASK_STATE_CHANGED, PERMISSION_GRANTED), optional AND-conjunctive filters, and harness target routing.
+      description: Subscribe / unsubscribe / update / list / resync / resume / rotate-key agent wake-event subscriptions with trigger types (SENT_TO_ME, TASK_STATE_CHANGED, PERMISSION_GRANTED), optional AND-conjunctive filters, and harness target routing.
       tags: [WakeSubscriptions]
       requestBody:
         required: true
@@ -2460,11 +2460,11 @@ paths:
               properties:
                 action:
                   type: string
-                  enum: [bootstrap, subscribe, unsubscribe, update, list, resync, resume]
+                  enum: [bootstrap, subscribe, unsubscribe, update, list, resync, resume, rotate-key]
                   description: "Lifecycle action to perform on the subscription. `resume` restores a route that was marked `status: degraded` after repeated delivery failures: it moves the persisted status back to `active` AND clears the in-flight markers in one step, which is what a restore requires - clearing only one half leaves the route skipped."
                 subscriptionId:
                   type: string
-                  description: Required for unsubscribe / update / resync / resume; optional for list (single).
+                  description: Required for unsubscribe / update / resync / resume / rotate-key; optional for list (single).
                 trigger:
                   type: string
                   enum: [SENT_TO_ME, TASK_STATE_CHANGED, PERMISSION_GRANTED]

--- a/ai/services/memory-core/WakeSubscriptionService.mjs
+++ b/ai/services/memory-core/WakeSubscriptionService.mjs
@@ -379,7 +379,7 @@ class WakeSubscriptionService extends Base {
      * action-specific handlers per ADR 0002 §6.6. ticket-ref-ok: decision-record authority, not issue archaeology
      *
      * @param {Object} opts
-     * @param {String} opts.action One of 'subscribe' | 'unsubscribe' | 'update' | 'list' | 'resync' | 'resume'
+     * @param {String} opts.action One of 'subscribe' | 'unsubscribe' | 'update' | 'list' | 'resync' | 'resume' | 'rotate-key'
      * @param {Object} [opts.rest] Action-specific parameters (see individual methods)
      * @returns {Promise<Object>}
      */
@@ -395,9 +395,10 @@ class WakeSubscriptionService extends Base {
             case 'list'       : return this.list       (rest);
             case 'resync'     : return this.resync     (rest);
             case 'resume'     : return this.resume     (rest);
+            case 'rotate-key' : return this.rotateKey  (rest);
             default:
                 throw new Error(
-                    `Invalid action '${action}'. Must be one of: bootstrap, subscribe, unsubscribe, update, list, resync, resume.`
+                    `Invalid action '${action}'. Must be one of: bootstrap, subscribe, unsubscribe, update, list, resync, resume, rotate-key.`
                 );
         }
     }
@@ -1139,6 +1140,74 @@ class WakeSubscriptionService extends Base {
         logger.info(`[WakeSubscription] resumed ${subscriptionId} for ${caller} (wasDegraded: ${wasDegraded})`);
 
         return {subscriptionId, status: 'active', wasDegraded};
+    }
+
+    /**
+     * @summary Re-issues the Shape-B signing key for a route that has lost one, in place.
+     *
+     * **Why this exists as its own action rather than inside `subscribe()`.** A row can hold an
+     * `a2a-webhook` target with no `signingKey`, and every prior repair door was closed: the mint runs
+     * only on `subscribe()`'s new-row branch, `subscribe()` on an existing route returns
+     * `{status: 'existing'}` before reaching it, and `update()` has no key surface. The only escape was
+     * unsubscribe + subscribe, which allocates a **new subscription id** — and that id is what the
+     * manifest, the receiver route table, delivery receipts, and the degrade all index on. Recovery by
+     * re-identification is not repair.
+     *
+     * **Why not mint inside `subscribe()`'s existing-row branch instead.** That call's contract is
+     * "ensure a row exists", and it is idempotent. Minting there would make every defensive
+     * re-subscribe a silent key rotation, and a *wrong* rotation is worse than the bug it fixes: it
+     * invalidates a `routes.json` a live receiver already validated, turning a deaf seat into a deaf
+     * seat plus a stale manifest. A caller survey settled it — every path into `subscribe()` is the
+     * `manage_wake_subscription` MCP tool, so a distinct action costs no existing caller anything.
+     *
+     * Server-issued only: the key is minted here and returned once, never accepted from the caller.
+     * The manifest generator fails closed on key disagreement, so an operator-supplied key would put
+     * two authorities on one secret.
+     *
+     * @param {Object} opts
+     * @param {String} opts.subscriptionId
+     * @returns {Promise<Object>} `{subscriptionId, signingKey, status: 'rotated', hadKey}`
+     */
+    async rotateKey({subscriptionId} = {}) {
+        const caller = RequestContextService.getAgentIdentityNodeId();
+        if (!caller) throw RequestContextService.unboundIdentityError('rotate signing key');
+        if (!subscriptionId) throw new Error("Missing 'subscriptionId' parameter.");
+
+        const subscription = this._loadSubscription(subscriptionId);
+        if (!subscription) throw new Error(`Subscription not found: ${subscriptionId}`);
+
+        // Owner-scoped, and this is the security boundary rather than a convenience: re-issuing a key
+        // for another seat's route hands the caller the ability to sign that seat's wakes.
+        if (subscription.agentIdentity !== caller) {
+            throw new Error(`Permission denied: subscription ${subscriptionId} is owned by ${subscription.agentIdentity}, not ${caller}.`);
+        }
+
+        if (subscription.harnessTarget !== 'a2a-webhook') {
+            throw new Error(
+                `Subscription ${subscriptionId} targets '${subscription.harnessTarget}', which carries no signing key. ` +
+                `Only a2a-webhook (Shape B) routes are signed.`
+            );
+        }
+
+        const
+            existingMetadata = subscription.harnessTargetMetadata || {},
+            hadKey           = Boolean(existingMetadata.signingKey),
+            signingKey       = crypto.randomBytes(32).toString('hex');
+
+        // Spread the stored metadata: `upsertNode` merges top-level properties, so replacing
+        // `harnessTargetMetadata` wholesale without the spread would drop `url` and the adapter tuple.
+        GraphService.upsertNode({
+            id        : subscriptionId,
+            properties: {harnessTargetMetadata: {...existingMetadata, signingKey}}
+        });
+
+        // The cache holds the pre-rotation record; a stale read would keep serving the old key.
+        this.subscriptionCache.delete(subscriptionId);
+
+        logger.info(`[WakeSubscription] rotated signing key for ${subscriptionId} (owner ${caller}, hadKey: ${hadKey})`);
+
+        // `hadKey` distinguishes repair from rotation for the caller, without logging either key.
+        return {subscriptionId, signingKey, status: 'rotated', hadKey};
     }
 
     /**

--- a/test/playwright/unit/ai/services/memory-core/WakeSubscriptionService.spec.mjs
+++ b/test/playwright/unit/ai/services/memory-core/WakeSubscriptionService.spec.mjs
@@ -2410,4 +2410,101 @@ test.describe('Neo.ai.services.memory-core.WakeSubscriptionService', () => {
             expect(res.idle).not.toContain('@neo-dispatch');
         });
     });
+
+    test.describe('rotateKey — the repair door for a row that lost its signing key', () => {
+        const shapeB = (overrides = {}) => insertDurableSubscription({
+            harnessTarget        : 'a2a-webhook',
+            harnessTargetMetadata: {url: 'http://127.0.0.1:3199/wake', adapter: 'osascript', appName: 'Claude'},
+            ...overrides
+        }).subscriptionId;
+
+        test('re-issues a key IN PLACE — the subscription id survives, which is the whole point', async () => {
+            // Recovery by unsubscribe+subscribe allocates a NEW id, and that id is what the manifest,
+            // the receiver route table, delivery receipts and the degrade all index on. Re-identification
+            // is not repair.
+            const id = shapeB();
+
+            await RequestContextService.run({agentIdentityNodeId: '@alice'}, async () => {
+                const result = await WakeSubscriptionService.rotateKey({subscriptionId: id});
+
+                expect(result.subscriptionId).toBe(id);
+                expect(result.status).toBe('rotated');
+                expect(result.hadKey).toBe(false);
+                expect(result.signingKey).toMatch(/^[0-9a-f]{64}$/);
+            });
+        });
+
+        test('preserves the rest of the metadata — a wholesale replace would drop the url', async () => {
+            const id = shapeB();
+
+            await RequestContextService.run({agentIdentityNodeId: '@alice'}, async () => {
+                await WakeSubscriptionService.rotateKey({subscriptionId: id});
+            });
+
+            const stored = GraphService.db.nodes.get(id).properties.harnessTargetMetadata;
+
+            expect(stored.url).toBe('http://127.0.0.1:3199/wake');
+            expect(stored.adapter).toBe('osascript');
+            expect(stored.appName).toBe('Claude');
+            expect(stored.signingKey).toMatch(/^[0-9a-f]{64}$/);
+        });
+
+        test('REFUSES a foreign owner — re-issuing another seat lets the caller sign that seat wakes', async () => {
+            const id = shapeB({owner: '@alice'});
+
+            await RequestContextService.run({agentIdentityNodeId: '@mallory'}, async () => {
+                await expect(WakeSubscriptionService.rotateKey({subscriptionId: id}))
+                    .rejects.toThrow(/Permission denied/);
+            });
+        });
+
+        test('the foreign attempt leaves no key behind — the refusal is before the mint', async () => {
+            const id = shapeB({owner: '@alice'});
+
+            await RequestContextService.run({agentIdentityNodeId: '@mallory'}, async () => {
+                await WakeSubscriptionService.rotateKey({subscriptionId: id}).catch(() => {});
+            });
+
+            expect(GraphService.db.nodes.get(id).properties.harnessTargetMetadata.signingKey).toBeUndefined();
+        });
+
+        test('rotates an EXISTING key too, and says which it did', async () => {
+            const id = shapeB({
+                harnessTargetMetadata: {url: 'http://127.0.0.1:3199/wake', signingKey: 'b'.repeat(64)}
+            });
+
+            await RequestContextService.run({agentIdentityNodeId: '@alice'}, async () => {
+                const result = await WakeSubscriptionService.rotateKey({subscriptionId: id});
+
+                expect(result.hadKey).toBe(true);
+                expect(result.signingKey).not.toBe('b'.repeat(64));
+            });
+        });
+
+        test('refuses a target that carries no key at all', async () => {
+            const id = insertDurableSubscription({harnessTarget: 'bridge-daemon'}).subscriptionId;
+
+            await RequestContextService.run({agentIdentityNodeId: '@alice'}, async () => {
+                await expect(WakeSubscriptionService.rotateKey({subscriptionId: id}))
+                    .rejects.toThrow(/carries no signing key/);
+            });
+        });
+
+        test('refuses an unbound identity rather than defaulting to an owner', async () => {
+            const id = shapeB();
+
+            await expect(WakeSubscriptionService.rotateKey({subscriptionId: id})).rejects.toThrow();
+        });
+
+        test('reaches the service through the MCP action name', async () => {
+            // A method with no tool-surface route is unreachable; the dispatch string is the contract.
+            const id = shapeB();
+
+            await RequestContextService.run({agentIdentityNodeId: '@alice'}, async () => {
+                const result = await WakeSubscriptionService.manage({action: 'rotate-key', subscriptionId: id});
+
+                expect(result.status).toBe('rotated');
+            });
+        });
+    });
 });

